File: unpruuf/app/src/main/java/com/nexonai/unpruuf/domain/relay/RelayPoolManager.kt
package com.nexonai.unpruuf.domain.relay

import android.content.Context
import android.util.Base64
import com.google.crypto.tink.subtle.Ed25519Verify
import com.nexonai.unpruuf.domain.network.RelayManager
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

/** The parsed, already-signature-verified contents of an imported relay-pool manifest. */
data class RelayPoolManifestInfo(
    val org: String,
    val issuedAtMs: Long,
    val relays: List<String>
)

/**
 * Offline import of a company-issued relay-pool manifest — see unpruuf/relaypool-tool/ for the
 * offline signer (keygen.js's private key never leaves that tool, and is a SEPARATE key pair
 * from LicenseManager's — a leaked relay-pool key only lets someone hand out bogus relay
 * ADDRESSES, never a valid app license, and vice versa). Format:
 * `unpruuf-relaypool:v1:<base64url(payload)>:<base64url(signature)>`, payload
 * `1|<org>|<issuedAtMs>|<relay1>;<relay2>;...` — same hand-rolled pipe-delimited style as
 * LicenseManager/RelayManager, verified with Tink's Ed25519Verify (cross-checked against
 * relaypool-tool's Node-side signer before shipping — a manifest signed with Node's own
 * `crypto.sign` verifies correctly here).
 *
 * Deliberately not device-bound and not gated behind any "enterprise account" concept — any
 * user who has the manifest text (received however the company distributes it: email, intranet,
 * printed card) can import it; see the plan's confirmed decision to keep v1 unrestricted. On
 * success, REPLACES (not merges) the locally configured backup relay pool via
 * [RelayManager.setExtraConnectionStrings] — re-importing an updated manifest is how a company
 * rotates its server list without asking every employee to hand-edit Settings.
 */
@Singleton
class RelayPoolManager @Inject constructor(
    @ApplicationContext private val context: Context,
    private val relayManager: RelayManager
) {
    private val prefs = context.getSharedPreferences("relay_pool_manifest", Context.MODE_PRIVATE)

    private val _lastImported = MutableStateFlow(loadStored())
    val lastImported: StateFlow<RelayPoolManifestInfo?> = _lastImported.asStateFlow()

    /** Verifies and applies [code]. Returns null on success, or a short user-facing reason on
     *  failure (nothing is changed in that case). */
    fun importManifest(code: String): String? {
        val info = parseAndVerify(code) ?: return "This relay-pool code isn't valid."
        relayManager.setExtraConnectionStrings(info.relays)
        prefs.edit()
            .putString(KEY_ORG, info.org)
            .putLong(KEY_ISSUED_AT, info.issuedAtMs)
            .putString(KEY_RELAYS, info.relays.joinToString(";"))
            .apply()
        _lastImported.value = info
        return null
    }

    private fun loadStored(): RelayPoolManifestInfo? {
        val org = prefs.getString(KEY_ORG, null) ?: return null
        val issuedAt = prefs.getLong(KEY_ISSUED_AT, -1L)
        val relays = RelayManager.parseConnectionStringList(prefs.getString(KEY_RELAYS, "") ?: "")
        if (issuedAt < 0 || relays.isEmpty()) return null
        return RelayPoolManifestInfo(org, issuedAt, relays)
    }

    private fun parseAndVerify(code: String): RelayPoolManifestInfo? {
        val trimmed = code.trim()
        if (!trimmed.startsWith(PREFIX)) return null
        val rest = trimmed.substring(PREFIX.length)
        val sep = rest.lastIndexOf(':')
        if (sep <= 0 || sep == rest.length - 1) return null
        val payload = decodeBase64Url(rest.substring(0, sep)) ?: return null
        val signature = decodeBase64Url(rest.substring(sep + 1)) ?: return null
        if (signature.size != 64) return null

        val publicKey = decodeBase64Url(PUBLIC_KEY_B64) ?: return null
        val verified = runCatching { Ed25519Verify(publicKey).verify(signature, payload) }.isSuccess
        if (!verified) return null

        val fields = String(payload, Charsets.UTF_8).split("|")
        if (fields.size != 4 || fields[0] != "1") return null
        val org = fields[1]
        if (org.isBlank()) return null
        val issuedAtMs = fields[2].toLongOrNull() ?: return null
        val relays = RelayManager.parseConnectionStringList(fields[3])
        if (relays.isEmpty() || relays.any { RelayManager.parseConnectionString(it) == null }) return null
        return RelayPoolManifestInfo(org, issuedAtMs, relays)
    }

    private fun decodeBase64Url(s: String): ByteArray? = runCatching {
        val rem = s.length % 4
        val padded = if (rem == 0) s else s + "====".substring(rem)
        Base64.decode(padded, Base64.URL_SAFE or Base64.NO_WRAP)
    }.getOrNull()

    companion object {
        private const val PREFIX = "unpruuf-relaypool:v1:"
        private const val KEY_ORG = "org"
        private const val KEY_ISSUED_AT = "issued_at_ms"
        private const val KEY_RELAYS = "relays"

        // Public half of a SEPARATE Ed25519 key pair from LicenseManager's — see
        // relaypool-tool/keygen.js. Safe to be public; only verifies, never signs. Must match
        // the constant in RelayPoolManager.swift on iOS exactly (both platforms verify manifests
        // issued by the same relaypool-tool key).
        const val PUBLIC_KEY_B64 = "jyPHi5C8bldr8B5p4lGz65Id3Rxqe0UFgov1A9r5y1A"
    }
}

File: unpruuf/ios/UnpruufApp/Sources/Services/RelayPoolManager.swift
import Combine
import Foundation
import UnpruufCore

/// App-level glue around `UnpruufCore.RelayPoolManifest`'s pure verifier — see that type's doc
/// comment for the wire format and cross-runtime verification story. This class owns *where* an
/// imported manifest is remembered (`UserDefaults`, same pattern `AppEnvironment` already uses
/// for `extraRelayConnectionStrings`). Applying a successfully-verified manifest to
/// `env.extraRelayConnectionStrings` is left to the caller (`SettingsView`) rather than done
/// here, so this type has no dependency on `AppEnvironment` — the same split already used there
/// for the hand-entered backup-relay field. Import is deliberately unrestricted (no device
/// binding, no "enterprise account" concept) — see the plan's confirmed v1 decision; any user
/// holding the manifest text can import it, and it REPLACES (not merges) the previous list.
@MainActor
final class RelayPoolManager: ObservableObject {
    /// Safe to be public; only verifies, never signs. Must match `RelayPoolManager.PUBLIC_KEY_B64`
    /// on Android exactly — both platforms verify manifests issued by the same relaypool-tool key.
    static let publicKeyB64 = "jyPHi5C8bldr8B5p4lGz65Id3Rxqe0UFgov1A9r5y1A"

    @Published private(set) var lastImported: RelayPoolManifest.Info?

    init() {
        if let org = UserDefaults.standard.string(forKey: Self.orgKey) {
            let issuedAtMs = UserDefaults.standard.object(forKey: Self.issuedAtKey) as? Int64 ?? -1
            let relays = RelayConnectionString.parseList(UserDefaults.standard.string(forKey: Self.relaysKey) ?? "")
            if issuedAtMs >= 0, !relays.isEmpty {
                self.lastImported = RelayPoolManifest.Info(org: org, issuedAtMs: issuedAtMs, relays: relays)
            }
        }
    }

    /// Verifies `code` and, on success, remembers it as the last-imported manifest (both in
    /// memory and in `UserDefaults`). Returns the parsed info on success — the caller is
    /// responsible for applying `info.relays` wherever it needs to go — or `nil` if `code` isn't
    /// a validly-signed manifest.
    func importManifest(_ code: String) -> RelayPoolManifest.Info? {
        guard let info = RelayPoolManifest.verify(code: code, publicKeyB64: Self.publicKeyB64) else {
            return nil
        }
        UserDefaults.standard.set(info.org, forKey: Self.orgKey)
        UserDefaults.standard.set(info.issuedAtMs, forKey: Self.issuedAtKey)
        UserDefaults.standard.set(RelayConnectionString.buildList(info.relays), forKey: Self.relaysKey)
        lastImported = info
        return info
    }

    private static let orgKey = "relay_pool_manifest_org"
    private static let issuedAtKey = "relay_pool_manifest_issued_at_ms"
    private static let relaysKey = "relay_pool_manifest_relays"
}
